File: 148.sort-list.py
#
# @lc app=leetcode id=148 lang=python3
#
# [148] Sort List
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def sortList(self, head: Optional[ListNode]) -> Optional[ListNode]:

        if not head or not head.next:
            return head
        
        slow, fast = head, head.next
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
        
        mid = slow.next
        slow.next = None

        left = self.sortList(head)
        right = self.sortList(mid)

        return self.merge_sort(left, right)



    def merge_sort(self, head_1, head_2):

        dummy_head = ListNode()
        curr_node = dummy_head

        while head_1 and head_2:
            if head_1.val < head_2.val:
                curr_node.next = head_1
                head_1 = head_1.next
            else:
                curr_node.next = head_2
                head_2 = head_2.next
            curr_node = curr_node.next
        
        if head_1:
            curr_node.next = head_1
        
        if head_2:
            curr_node.next = head_2
        
        return dummy_head.next
        


        
# @lc code=end

File: 148.sort-list.java
/*
 * @lc app=leetcode id=148 lang=java
 *
 * [148] Sort List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode sortList(ListNode head) {
        
        if (head == null || head.next == null) {
            return head;
        }

        ListNode slow = head;
        ListNode fast = head.next;

        while(fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
        }

        ListNode mid = slow.next;
        slow.next = null;

        ListNode left = sortList(head);
        ListNode right = sortList(mid);

        return mergeSort(left, right);

    }

    private ListNode mergeSort(ListNode left, ListNode right) {

        ListNode dummyHead = new ListNode();
        ListNode currentNode = dummyHead;

        while (left != null && right != null) {
            
            if (left.val < right.val) {
                currentNode.next = left;
                left = left.next;
            } else {
                currentNode.next = right;
                right = right.next;
            }

            currentNode = currentNode.next;
        }

        if (left != null) {
            currentNode.next = left;
        }

        if (right != null) {
            currentNode.next = right;
        }

        return dummyHead.next;
    }
}
// @lc code=end

